### DISCUSSION TITLE
Why discrete pulldown resistor on RPI input for flow switch?

### DISCUSSION THREAD

**Discussion by davthomaspilot** [tagyoureit/nodejs-poolController#735] [REM] [answered]
Versus configuring the input with a pulldown in software.

(I have a great selection of 0805 and 0603 SMT resistors.  Axial resistors, not so much.)

**Comment by rstrouse** [comment]
No reason other than it s more difficult to have folks modify the pin configs in the os.

**Reply by davthomaspilot** [reply]
Actually, I think all the GPIO default to inputs, with either pull-up or pull-down.  Here's a table showing which are pulled up versus pulled down:

![image](https://user-images.githubusercontent.com/6830886/233874344-e91ce8fe-e1bb-44c4-89f8-43f621bfd895.png)

So, you just need to pick a GPIO which defaults low.  

The picture you showed uses GPIO 5 (pin  which defaults to pull-up high.  So, that pin has internal pull-up of 50 - 65K and the external pull-down of 10K.  Level will be about 1/6 x 3.3 = .5 V.

That's still a good logic low, but given the wire running to the flow switch it would be better to turn-off the pull-up using pigs, or simply choose a GPIO that defaults to pull down and loose the discrete resistor.

**Reply by rstrouse** [reply, marked as answer]
You can do it either way.  The docs were simply created to be as prescriptive as possible so folks are not confused about all the BCM nuttiness.  That chart also gets blown up if settings are changed in the pin configuration.  I also make sure I put a protection resistor on pins simply because I don't want a dead short into the pin to torch it.  Just preference for me and it keeps me honest.

**Reply by davthomaspilot** [reply]
"I also make sure I put a protection resistor on pins simply because I don't want a dead short into the pin to torch it. "

That makes sense if you drive one side of the relay straight from 3.3V.  

But, GPIO outputs are short-circuit protected and you can define the drive strength.  So, you can eliminate the need  for any external resistors by driving that side of the relay from a GPIO output rather than 3.3V.

REM already requires the user to associate a a GPIO input with one of the flow sensor wires.  You could also have an output associated with the other wire of the switch, rather than assuming it's hardwired to 3.3V through a resistor.

I can see someone using crimped pin connectors instead of soldering the resistors.  If the crimp is marginal, the connector comes off or corrodes over time, the flow sensor could have false "flow true" indication.  With the dosing pumps wired with "always on" 24VDC, dosing would occur even if the pool pumps are off--unless the code "knows" when the pool pump is off and prevents dosing that way too.

I'm all about eliminating any unnecessary hardware, but to each his own.

**Reply by davthomaspilot** [reply]
I just noticed there are is a place for a 3 pin header on the Whitebox tentacle MKII:

![wpMKII](https://user-images.githubusercontent.com/6830886/234098171-edbee05f-40e2-4fa5-91a1-0a29404e0dd1.png)

There's a silkscreen box that's labelled "flow".

The pins are labelled R, W, and B.  

I can't find the schematics for the MKII version--it's not there on the version of the schematic I've been able to find online.

Be nice if I could just add a JST style connector there....

**Reply by Telemachus1977** [reply]
> Actually, I think all the GPIO default to inputs, with either pull-up or pull-down. Here's a table showing which are pulled up versus pulled down:
> 
> ![image](https://user-images.githubusercontent.com/6830886/233874344-e91ce8fe-e1bb-44c4-89f8-43f621bfd895.png)
> 
> So, you just need to pick a GPIO which defaults low.
> 
> The picture you showed uses GPIO 5 (pin which defaults to pull-up high. So, that pin has internal pull-up of 50 - 65K and the external pull-down of 10K. Level will be about 1/6 x 3.3 = .5 V.
> 
> That's still a good logic low, but given the wire running to the flow switch it would be better to turn-off the pull-up using pigs, or simply choose a GPIO that defaults to pull down and loose the discrete resistor.

Would be great if the GPIO pins in REM were labeled or colored differently by high/low input.

This is a great table!

**Comment by davthomaspilot** [comment]
Turns out that's where you can plug the EZO-flow totalizer.

There's no support for that in REM, is there?

**Comment by rstrouse** [comment]
I am not sure but I think those are alarm pins for Pressure and Humidity.  They are likely wired directly to pin 5 on the two external headers.  I can't be sure because I do not have the MKII.

**Reply by davthomaspilot** [reply]
Just following up--

R  -  3.3V
W - PGND  (the GND of the non-isolated sensor
B   - PRB

The "R" hold is a nice place to solder the 220 ohm resistor to  3.3V.  There's room to lay it flat on the pcb and strain relieve with hot melt glue or goop.

Yeah, I know I was arguing to lose the resistor, but this is mechanically better than soldering to the GPIO header.

I still need to pick up a "normally low" input GPIO from the RPI header.
